src/SurfaceFluxes.jl

```diff
@@ -1,399 +1,519 @@
-    𝓁u₀ = compute_z0(u★₀, param_set, sc, sc.roughness_model, UF.MomentumTransport())
-    𝓁θ₀ = compute_z0(u★₀, param_set, sc, sc.roughness_model, UF.HeatTransport())
-    𝓁q₀ = compute_z0(u★₀, param_set, sc, sc.roughness_model, UF.HeatTransport())
-    # Initial guesses for MOST iterative solution
-    if ΔDSEᵥ(param_set, sc) >= FT(0)
-        X★₀ = (u★ = u★₀, DSEᵥ★ = FT(δ), θᵥ★ = FT(δ), q★ = FT(δ),
-            L★ = FT(10),
-            𝓁u = 𝓁u₀, 𝓁θ = 𝓁θ₀, 𝓁q = 𝓁q₀)
-        X★ = obukhov_iteration(X★₀, sc, scheme, param_set, tol)
-        return X★
-    else
-        X★₀ = (u★ = u★₀, DSEᵥ★ = FT(δ), θᵥ★ = FT(δ), q★ = FT(δ),
-            L★ = FT(-10),
-            𝓁u = 𝓁u₀, 𝓁θ = 𝓁θ₀, 𝓁q = 𝓁q₀)
-        X★ = obukhov_iteration(X★₀, sc, scheme, param_set, tol)
-        return X★
-    end
+    # Momentum fluxes
+    gustiness = gustiness_value(inputs.gustiness_model, param_set, b_flux)
+    (ρτxz, ρτyz) = momentum_fluxes(Cd, inputs, ρ_sfc, gustiness)
+
+    return (shf, lhf, E, ρτxz, ρτyz)
 end
 
-function obukhov_similarity_solution(
-    param_set,
-    sc::FluxesAndFrictionVelocity,
-    scheme,
-    args...,
-)
-    return (L★ = -sc.ustar^3 / SFP.von_karman_const(param_set) /
-                 non_zero(compute_buoyancy_flux(param_set, sc, scheme)), u★ = sc.ustar)
+struct ResidualFunction{PS, I, UF, TP, SCH} <: Function
+    param_set::PS
+    inputs::I
+    scheme::SCH
+    uf_params::UF
+    thermo_params::TP
 end
 
-"""
-    momentum_fluxes(param_set, Cd, sc, scheme)
-
-Compute and return the momentum fluxes
-## Arguments
-  - param_set: Abstract Parameter Set containing physical, thermodynamic parameters.
-  - Cd: Momentum exchange coefficient
-  - sc: Container for surface conditions based on known combination
-        of the state vector, and {fluxes, friction velocity, exchange coefficients} for a given experiment
-  - scheme: Discretization scheme (currently supports FD and FV)
-"""
-function momentum_fluxes(param_set, Cd, sc::AbstractSurfaceConditions, scheme)
-    thermo_params = SFP.thermodynamics_params(param_set)
-    ρ_sfc = TD.air_density(thermo_params, ts_sfc(sc))
-    ρτxz = -ρ_sfc * Cd * Δu1(sc) * windspeed(sc)
-    ρτyz = -ρ_sfc * Cd * Δu2(sc) * windspeed(sc)
-    return (ρτxz, ρτyz)
+function (rf::ResidualFunction)(ζ)
+    FT = eltype(rf.param_set)
+
+    # Unpack parameters that do not change over iterations
+    param_set = rf.param_set
+    inputs = rf.inputs
+    scheme = rf.scheme
+    uf_params = rf.uf_params
+    thermo_params = rf.thermo_params
+
+    # 1. Compute u_star and roughness lengths, iteratively if they are mutually dependent
+    u_star, z0m, z0s = compute_ustar_and_roughness(
+        param_set,
+        ζ,
+        inputs,
+        scheme,
+    )
+    z0h = z0s # Assuming scalar roughness = heat roughness
+
+    # 2. Update T_sfc and q_vap_sfc via callbacks or use inputs
+    T_sfc_new = eval_callback(inputs.update_T_sfc, inputs.T_sfc_guess, ζ, param_set, thermo_params, inputs, FT)
+    q_vap_sfc_new =
```

Comment:
I think a use case is to solve for T_sfc_new and then say q_sfc_new = q_sat(T_sfc_new), or some function of q_sat(T_sfc_new). Say for sea ice or snow. Im not sure if we can do that here unless the update_q_vap_sfc function also calls update_T_sfc internally, or if we also pass T_sfc as an argument somehow

Reply:
maybe it could use rho_int when computing q_sat instead of rho_sfc, since rho_int is already in `inputs`, and rho_sfc has to be computed

Reply:
I added T_sfc as an input.